src/requests_gssapi/gssapi_.py

```diff
@@ -144,6 +152,37 @@ def generate_request_header(self, response, host, is_preemptive=False):
         if self.mutual_authentication != DISABLED:
             gssflags.append(gssapi.RequirementFlag.mutual_authentication)
 
+        gss_cb = None
+        if self.channel_bindings == "tls-server-end-point":
+            if is_preemptive:
+                log.warning("channel_bindings were requested, but are unavailable for opportunistic authentication")
+            # The 'connection' attribute on raw is a public urllib3 API
+            # and can be None if the connection has been released.
+            elif getattr(response.raw, "connection", None) and getattr(response.raw.connection, "sock", None):
+                try:
+                    # Defer import so it's not a hard dependency.
+                    from cryptography import x509
+
+                    sock = response.raw.connection.sock
+
+                    der_cert = sock.getpeercert(binary_form=True)
+                    cert = x509.load_der_x509_certificate(der_cert)
+                    hash = cert.signature_hash_algorithm
+                    cert_hash = cert.fingerprint(hash)
+
+                    app_data = b"tls-server-end-point:" + cert_hash
+                    gss_cb = gssapi.ChannelBindings(application_data=app_data)
+                    log.debug("generate_request_header(): Successfully retrieved channel bindings")
+                except ImportError:
+                    log.warning("Could not import cryptography, python-cryptography is required for this feature.")
+                except Exception:
+                    log.warning(
+                        "Failed to get channel bindings from socket",
+                        exc_info=True,
+                    )
+            else:
+                log.warning("channel_bindings were requested, but a socket could not be retrieved from the response")
```

Comment:
Do we want any of these to be a warning, I would have thought if someone explicitly requested the CB and we failed then they would want to know.

Reply:
I think those should be errors.

Reply:
ok should I simply remove the try/except and let any error bubble up ?